// tiktok-info/index.js
const cheerio = require('cheerio')
const needle = require('needle');

exports.download = async(url) => {
    const response = await needle('get', url, null, {
        headers: {
            user_agent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.34',
            accept: 'text/html,application/xhtml+xml'
        },
        follow_max: 5
    });

    const $ = await cheerio.load(response.body);
    let scripts = $('script').get();
    let data = undefined;
    for (let s of scripts) {
        if (s.children[0] && s.children[0].data && s.children[0].data.startsWith("window.__INIT_PROPS__")) {
            const raw = s.children[0].data;
            const json = raw.replace(/window.__INIT_PROPS__ ?= ?/, "");
            data = JSON.parse(json);
        }

    }
    const videoData = data['/v/:id'].videoData;
    const author = videoData.authorInfos;
    const videoUrl = videoData.itemInfos.video.urls[0];
    if (!videoUrl) {
        return Promise.reject({ status: 'video-not-found' });
    }

    const description = videoData.itemInfos.text;

    const video = {
        status: 'ok',
        author: author,
        description,
        videoUrl,
    }

    return Promise.resolve(video);
};
